#### Ryan Gilbert
#### CMP 262-56145
#### Final Project - Data Cleaning

# Objectives

## Determine Answers to Following and Resolve: 

### - Did your exploratory analysis from last week highlight anything that needs to be cleaned?
### - Investigate any NaN's that you find. Should these be replaced with another value? If so, what would be the best value? Should that record be deleted instead?
### - Are there outliers to any of the data items? What could explain those outliers? Should those records be changed or deleted?
### - Are the column names not in a format that makes it easy to refer to?
### - Do you need to merge multiple datasets into one?
### - Do you need to split up a column's data into multiple columns or simplify its values?

# Import Modules for Data Cleaning

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Using Pandas to Read CSV File & Saving to Pandas Dataframe

In [2]:
df = pd.read_csv("ISPP_School_Summary.csv")

### Confirming that Dataset was Successfully Extracted

In [3]:
df.sample()

Reporting_Year  School_Code                 School_Name  District_Code  \
842      2022-2023     47250109  Newton Senior High  School           4725   

                   District_Name       AEA_Name       Grade_Range School_Type  \
842  Newton Comm School District  Heartland AEA  (09, 10, 11, 12)        High   

     School_Index School_Rating_Category  ... Index_EL Index_FRL Index_IEP  \
842         54.42             Acceptable  ...      NaN     50.56     43.47   

    Index_Asian Index_Black Index_PacIs Index_Hispanic Index_Multi  \
842         NaN         NaN         NaN          52.13         NaN   

    Index_NatAm Index_White  
842         NaN       54.53  

[1 rows x 71 columns]

#### Noting that there are 71 columns in the original data frame

# Cleaning Columns that have Missing Values

### Confirming column names

In [4]:
df.columns

Index(['Reporting_Year', 'School_Code', 'School_Name', 'District_Code',
       'District_Name', 'AEA_Name', 'Grade_Range', 'School_Type',
       'School_Index', 'School_Rating_Category', 'ESSA_Support_Status',
       'Comprehensive_Decision', 'Targeted_Decision', 'Targeted_Subgroups',
       'School_Index_Source', 'Address1', 'Address2', 'City', 'State', 'ZIP',
       'Phone', 'Fax', 'Website', 'Email', 'School_Administrator_Name',
       'School_Administrator_Title', 'Count_Male', 'Percent_Male',
       'Count_Female', 'Percent_Female', 'Count_EL', 'Percent_EL', 'Count_FRL',
       'Percent_FRL', 'Count_IEP', 'Percent_IEP', 'Count_Asian',
       'Percent_Asian', 'Count_Black', 'Percent_Black', 'Count_PacIs',
       'Percent_PacIs', 'Count_Hispanic', 'Percent_Hispanic', 'Count_Multi',
       'Percent_Multi', 'Count_NatAm', 'Percent_NatAm', 'Count_White',
       'Percent_White', 'Count_Foster_Care', 'Percent_Foster_Care',
       'Count_Military_Connected', 'Percent_Military_Connected',


### From the Data Exploration, the first columns that will be removed are 'Address2', 'Website', and 'Phone'.  

In [5]:
df_updated_cols = df.drop(['Address2', 'Website', 'Phone', 'Fax'], axis = 1)

#### Expecting there to be 67 Remaining Columns after removing 3 from the original 71

In [6]:
df_updated_cols.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1295 entries, 0 to 1294
Data columns (total 67 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Reporting_Year              1295 non-null   object 
 1   School_Code                 1295 non-null   int64  
 2   School_Name                 1295 non-null   object 
 3   District_Code               1295 non-null   int64  
 4   District_Name               1295 non-null   object 
 5   AEA_Name                    1295 non-null   object 
 6   Grade_Range                 1295 non-null   object 
 7   School_Type                 1295 non-null   object 
 8   School_Index                1295 non-null   float64
 9   School_Rating_Category      1295 non-null   object 
 10  ESSA_Support_Status         1295 non-null   object 
 11  Comprehensive_Decision      1295 non-null   object 
 12  Targeted_Decision           1295 non-null   object 
 13  Targeted_Subgroups          348 n

#### 67 Columns retained confirmed in original dataframe. 

### The Remaining Columns with missing contain information which may be relevant for Data Analysis. Those remaining columns include: Target_Subgroups, School_Index_Source, Index_EL, Index_FRL, Index_IEP, _Index_Asian, Index_Black_Index_PacIS, _Index_hispanic, Index_Multi, Index_NatAm, Index_white

### Based on Documentation with the Dataset, the Target_Subgroups Column is used when the school is designated as Targeted (under ESSA) and provides the list of subgroups that have an accountability index score at or below 'Comprehensive' score. The field is left blank when the school is not 'Target' for ESSA support, and when if it was Targeted, that it met the 'Comprehensive' index score in all of its student accountability groups. Therefore, 'Achieved Standard' would be appropriate for these blank cells.

### The School Index Source is used when there are too few students enrolled in a school to achieve a calculated accountability index score. This column lists all schools that it inherits its score from, so when it is blank, that indicates that the school is large enough to account for its own data. So, in this case, the blank cells can be filled with 'Independent Institution Data'.

### The remaining columns with missing values are all a bit more straight forward. They represent an index score for a specific subgroup at that school. If it is blank, the subgroup is not represented enough to have an accountability index score. The maximum value for these indexes is 100 and the minimum score would be 1. Therefore, a value of '0' will be added to each of these missing values instead.

In [7]:
df_clean = df_updated_cols 

### Updating Index Scores

#### Creating the list of Index Scores so that they can have all 'NAN' cells filled with '0'

In [8]:
Index_list = ['Index_EL', 'Index_FRL', 'Index_IEP', 'Index_Asian', 'Index_Black', 'Index_PacIs', 'Index_Hispanic', 'Index_Multi', 'Index_NatAm', 'Index_White']

In [9]:
df_clean[Index_list] = df_clean[Index_list].fillna(0)

#### Checking to make sure that the step was successful

In [10]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1295 entries, 0 to 1294
Data columns (total 67 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Reporting_Year              1295 non-null   object 
 1   School_Code                 1295 non-null   int64  
 2   School_Name                 1295 non-null   object 
 3   District_Code               1295 non-null   int64  
 4   District_Name               1295 non-null   object 
 5   AEA_Name                    1295 non-null   object 
 6   Grade_Range                 1295 non-null   object 
 7   School_Type                 1295 non-null   object 
 8   School_Index                1295 non-null   float64
 9   School_Rating_Category      1295 non-null   object 
 10  ESSA_Support_Status         1295 non-null   object 
 11  Comprehensive_Decision      1295 non-null   object 
 12  Targeted_Decision           1295 non-null   object 
 13  Targeted_Subgroups          348 n

#### All 'Index' subgroup columns now have 1295 values and therefore no missing values. 

## Updating Target_Subgroups and then School_Index_Source Based on Info Above. 

In [11]:
df_clean['Targeted_Subgroups'] = df_clean['Targeted_Subgroups'].fillna('Achieved Standard')

In [12]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1295 entries, 0 to 1294
Data columns (total 67 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Reporting_Year              1295 non-null   object 
 1   School_Code                 1295 non-null   int64  
 2   School_Name                 1295 non-null   object 
 3   District_Code               1295 non-null   int64  
 4   District_Name               1295 non-null   object 
 5   AEA_Name                    1295 non-null   object 
 6   Grade_Range                 1295 non-null   object 
 7   School_Type                 1295 non-null   object 
 8   School_Index                1295 non-null   float64
 9   School_Rating_Category      1295 non-null   object 
 10  ESSA_Support_Status         1295 non-null   object 
 11  Comprehensive_Decision      1295 non-null   object 
 12  Targeted_Decision           1295 non-null   object 
 13  Targeted_Subgroups          1295 

In [13]:
df_clean['School_Index_Source'] = df_clean['School_Index_Source'].fillna('Independent Institutional Data')

In [14]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1295 entries, 0 to 1294
Data columns (total 67 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Reporting_Year              1295 non-null   object 
 1   School_Code                 1295 non-null   int64  
 2   School_Name                 1295 non-null   object 
 3   District_Code               1295 non-null   int64  
 4   District_Name               1295 non-null   object 
 5   AEA_Name                    1295 non-null   object 
 6   Grade_Range                 1295 non-null   object 
 7   School_Type                 1295 non-null   object 
 8   School_Index                1295 non-null   float64
 9   School_Rating_Category      1295 non-null   object 
 10  ESSA_Support_Status         1295 non-null   object 
 11  Comprehensive_Decision      1295 non-null   object 
 12  Targeted_Decision           1295 non-null   object 
 13  Targeted_Subgroups          1295 

#### All remaining columns have 1295 non-null values, meaning there are no missing values remaining in the DataFrame. Next, some additional columns will be removed from the Data Frame that will not be used to answer the Data Questions.

### Columns that do not include data that can be used to answer data questions include the Reporting Year (all data collected was from the same Reporting Year), all School Addres Information, the Grade Range (in this case School Type captures same information), the AEA Name, and the Counts for the Subgroups since Percentages will be more appropriate to compare schools against each other. If a school does not have enough students (either in total) or of a subgroup, they will not have their own Index Score, so the counts will be unnecessary information.

In [15]:
df_final = df_clean.drop(['Reporting_Year', 'AEA_Name', 'Grade_Range', 'Address1', 'City', 'State', 'ZIP', 'Email', 'School_Administrator_Name', 'School_Administrator_Title'], axis=1)

#### Removed 10 total columns, expecting 57 remaining (were 67)

In [16]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1295 entries, 0 to 1294
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   School_Code                 1295 non-null   int64  
 1   School_Name                 1295 non-null   object 
 2   District_Code               1295 non-null   int64  
 3   District_Name               1295 non-null   object 
 4   School_Type                 1295 non-null   object 
 5   School_Index                1295 non-null   float64
 6   School_Rating_Category      1295 non-null   object 
 7   ESSA_Support_Status         1295 non-null   object 
 8   Comprehensive_Decision      1295 non-null   object 
 9   Targeted_Decision           1295 non-null   object 
 10  Targeted_Subgroups          1295 non-null   object 
 11  School_Index_Source         1295 non-null   object 
 12  Count_Male                  1295 non-null   int64  
 13  Percent_Male                1295 

In [17]:
Count_List = ['Count_Male', 'Count_Female', 'Count_EL', 'Count_FRL', 'Count_IEP', 'Count_Asian', 'Count_Black', 'Count_PacIs', 'Count_Hispanic', 'Count_Multi', 'Count_NatAm', 'Count_White', 'Count_Foster_Care', 'Count_Military_Connected', 'Count_Migrant', 'Count_Homeless', 'Count_NonBinary'] 

In [18]:
Count_List

['Count_Male',
 'Count_Female',
 'Count_EL',
 'Count_FRL',
 'Count_IEP',
 'Count_Asian',
 'Count_Black',
 'Count_PacIs',
 'Count_Hispanic',
 'Count_Multi',
 'Count_NatAm',
 'Count_White',
 'Count_Foster_Care',
 'Count_Military_Connected',
 'Count_Migrant',
 'Count_Homeless',
 'Count_NonBinary']

#### This list of 17 Counts should bring the total number of columns down to 40 from 57

In [19]:
df_final = df_final.drop(Count_List, axis = 1)

In [20]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1295 entries, 0 to 1294
Data columns (total 40 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   School_Code                 1295 non-null   int64  
 1   School_Name                 1295 non-null   object 
 2   District_Code               1295 non-null   int64  
 3   District_Name               1295 non-null   object 
 4   School_Type                 1295 non-null   object 
 5   School_Index                1295 non-null   float64
 6   School_Rating_Category      1295 non-null   object 
 7   ESSA_Support_Status         1295 non-null   object 
 8   Comprehensive_Decision      1295 non-null   object 
 9   Targeted_Decision           1295 non-null   object 
 10  Targeted_Subgroups          1295 non-null   object 
 11  School_Index_Source         1295 non-null   object 
 12  Percent_Male                1295 non-null   float64
 13  Percent_Female              1295 

#### As expected, the total number of columns was brought to 40 from 57. For individual questions, new data-frames can be created using the remaining columns from this cleaned DataFrame.

### All of the remaining columns are in the appropriate data type: Percentages are Float, the remaining Count for the Total Students is an integer, and the names of schools and statuses are all Objects.

### With regards to Outliers in the data-set, despite utilizing the 'describe()' function, the remaining columns with numeric data include values such as School Code and District Code, where variation would be expected, as well as Index Scores, which are calculated based on the Schools' Performances and should not be modified as well as Percentages of different subgroups within the school, which again should not be modified or removed, as having anomalies in either will assist in finding trends in the data to answer some of the data questions posed. 

### Column names are all concise and intuitive, so no modifications of those names have been made, especially now that all columns that will not be used to answer Data Questions have been removed from the original Dataset.

### Finally with regards to merging and splitting data, that is unnecessary with this dataset, as the entire Iowa School Performance Profile data collected for the 2022-2023 Academic Reporting Year, rather than all historic data ever collected since the inititative began, and the remaining data columns are all simplified values. 

# Final Product

In [21]:
print(df_final)

      School_Code                             School_Name  District_Code  \
0           90109                       AGWSR High School              9   
1           90172                     AGWSR Middle School              9   
2           90409                 AGWSR Elementary School              9   
3           90427                         The Cougars Den              9   
4           90436           AGWSR Elementary at Wellsburg              9   
...           ...                                     ...            ...   
1290     71100118                        Woodward Academy           7110   
1291     71100409          Woodward-Granger Middle School           7110   
1292     71100418      Woodward-Granger Elementary School           7110   
1293     71100427  Woodward-Granger Early Learning Center           7110   
1294     81000101                   Choice Charter School           8100   

                              District_Name            School_Type  \
0                

## Final: All 1295 Rows Remain from the original data set and 71 columns has been reduced to 40. There are no remaining 'NAN' values remaining in any of the dataset, and all of the remaining data includes consistent and simplified data types.

## Save Cleaned Data Frame so that it can be Accessed by Data Results & Visualizations Notebook

In [22]:
df_final.to_csv("df_final.csv", index=False)